### DISCUSSION TITLE
Redux usage in React Flow

### DISCUSSION THREAD

**Discussion by mueen-memon** [xyflow/xyflow#2126] [Q&A] [answered]
Hi there, 
I am trying to use redux in react flow.
I can create nodes and edges while using drag and drop. The problem is I am unable to move them inside the canvas.  I think for controlled Flow I need to use applyNodeChanges but I am unable to use it properly.

**Comment by bcakmakoglu** [comment, marked as answer]
Can you provide a minimal example of your code?

I'm just guessing here but did you pass a cb to `onNodesChange` / `onEdgesChange` ?

```
 onNodesChange={onNodesChange}
 onEdgesChange={onEdgesChange}
```

If you're using a separate state you would have to apply the changes to them so something like that
```
const onNodesChange = (changes) => {
  setMyStoredNodes(applyNodeChanges(changes, myStoredNodes))
}
```

[Here's](https://codesandbox.io/s/sandpack-project-forked-n3085t?file=/App.js:682-1422) an example of what I mean, maybe this will help you out.

**Reply by mueen-memon** [reply]
Thanks.
